lib/bolt/pal/yaml_plan/evaluator.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+require 'bolt/pal/yaml_plan'
+
+module Bolt
+  class PAL
+    class YamlPlan
+      class Evaluator
+        def initialize
+          @logger = Logging.logger[self]
+          @evaluator = Puppet::Pops::Parser::EvaluatingParser.new
+        end
+
+        STEP_KEYS = %w[task command eval].freeze
+
+        def dispatch_step(scope, step)
+          step = evaluate_code_blocks(scope, step)
+
+          step_type, *extra_keys = STEP_KEYS.select { |key| step.key?(key) }
+          if !step_type || extra_keys.any?
+            unsupported_step(scope, step)
+          end
+
+          case step_type
+          when 'task'
+            task_step(scope, step)
+          when 'command'
+            command_step(scope, step)
+          when 'eval'
+            eval_step(scope, step)
+          else
+            # This shouldn't be able to happen since this case statement should
+            # match the STEP_KEYS list, but raise an error *just in case*,
+            # instead of silently skipping the step.
+            unsupported_step(scope, step)
+          end
+        end
+
+        def task_step(scope, step)
+          task = step['task']
+          target = step['target']
+          description = step['description']
+          params = step['parameters'] || {}
+          raise "Can't run a task without specifying a target" unless target
+
+          args = if description
+                   [task, target, description, params]
+                 else
+                   [task, target, params]
+                 end
+          scope.call_function('run_task', args)
+        end
+
+        def command_step(scope, step)
+          command = step['command']
+          target = step['target']
+          description = step['description']
+          raise "Can't run a command without specifying a target" unless target
+
+          args = [command, target]
+          args << description if description
+          scope.call_function('run_command', args)
+        end
+
+        def eval_step(_scope, step)
+          step['eval']
```

Comment:
In order to bring this up to the level of the other steps I think we should add some logging.

Reply:
What sort of logging do you think we should have here?

It's not entirely straightforward to log like "starting eval/finished eval" because of the current execution model. The code is actually eval-ed _before_ we ever call `eval_step`, since `dispatch_step` evaluates all the nested code blocks before dispatching.

We could log something like "eval-ed {code block} with result {foo}" though.

Reply:
Actually, by that point we will have thrown out the original code and replaced it with its result.

The behavior of this is fairly unintuitive and really might need to be handled inside the `eval_step` method. It's really weird that `eval: notice("hello world")` doesn't work. Because it's a bare string and doesn't start with a `$`, it's not treated as code. That seems fine in most cases, but very wrong with `eval`.

I think there's a middle ground between the current behavior where `eval` is _just_ a way to assign the value of an expression to a variable, and the possibility where `eval` _always_ treats its argument as code, resulting in double evaluation. Because we have the marker class `BareString`, we could hold off on evaluating the `eval` value until we get into the `eval_step` method, and reinterpret any `BareString` as a `CodeLiteral` instead, so that it will always be evaluated.

That behavior is probably more intuitive, but also quite difficult to explain in detail...

Reply:
well you've at least convinced me to wait to discuss this until we can do some user testing.